### PR TITLE
Minimal implementation of QImage.

### PR DESCRIPTION
**Pull request by LeonMatthesKDAB**
Just enough to make the qt-tutorial for KDAB Training day work.

### CODE REVIEW

File: crates/cxx-qt-lib/src/core/qbytearray.rs [hunk 1]
```diff
@@ -122,6 +125,97 @@ impl AsRef<[u8]> for QByteArray {
     }
 }
 
+/// A QByteArray cursor allows seeking inside a QByteArray
+/// and writing to it.
+/// Comparable with std::io::Cursor.
+pub struct QByteArrayCursor {
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-lib/src/core/qbytearray.rs:131]: can we split the QByteArray changes into a separate commit?

Comment (LeonMatthesKDAB) [reply crates/cxx-qt-lib/src/core/qbytearray.rs:131]: Done

File: crates/cxx-qt-lib/src/core/qbytearray.rs [hunk 2]
```diff
@@ -122,6 +125,97 @@ impl AsRef<[u8]> for QByteArray {
...
+        }
+    }
+
+    /// Returns a reference to the underlying QByteArray.
+    pub fn into_inner(self) -> QByteArray {
+        self.inner
+    }
+}
+
+/// Like the implementation of Cursor<Vec<u8>>, the Cursor<QByteArray>
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-lib/src/core/qbytearray.rs:153]: do this doc go on the `impl` or on the `fn` ? does it work in a cargo co ?

File: crates/cxx-qt-lib/src/gui/qimage.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+use crate::QByteArray;
+use cxx::{type_id, ExternType};
+use std::{
+    ffi::{c_char, CString},
+    mem::MaybeUninit,
+};
+
+#[cxx::bridge]
+mod ffi {
+
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-lib/src/gui/qimage.rs:15]: random empty line :-)

File: crates/cxx-qt-lib/src/gui/qimage.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        fn drop(image: &mut QImage);
+
+        #[doc(hidden)]
+        #[rust_name = "qimage_init_from_data"]
+        unsafe fn qimageInitFromData(data: &QByteArray, format: *const c_char) -> QImage;
+    }
+}
+
+#[repr(C)]
+pub struct QImage {
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-lib/src/gui/qimage.rs:42]: needs docs, but as discussed needs to stated it needs to be marked as experimental / WIP etc

File: crates/cxx-qt-lib/src/gui/qimage.cpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                          sizeof(::std::size_t) * 3);
+#endif
+
+namespace rust {
+namespace cxxqtlib1 {
+
+QImage
+qimageInitFromData(const QByteArray& data, const char* format)
+{
+  return QImage::fromData(data, format);
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-lib/src/gui/qimage.cpp:41]: as we need a C++ proxy anyway, why not pass in `rust::Str` here https://cxx.rs/binding/str.html then this would avoid the `unsafe` method `qimageInitFromData` ?

File: crates/cxx-qt-lib/src/gui/qimage.rs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+impl Drop for QImage {
+    fn drop(&mut self) {
+        ffi::qimage_drop(self);
+    }
+}
+
+impl QImage {
+    pub fn from_data(data: &QByteArray, format: &str) -> Option<Self> {
+        let format = CString::new(format).ok()?;
+        let image = unsafe { ffi::qimage_init_from_data(data, format.as_ptr() as *const c_char) };
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-lib/src/gui/qimage.rs:68]: then here you just pass in the `format` and you don't need to convert to a `CString`

File: crates/cxx-qt-lib/src/gui/qimage.rs (hunk 3 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-lib/src/gui/qimage.rs:15]: Done

File: crates/cxx-qt-lib/src/gui/qimage.cpp (hunk 5 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-lib/src/gui/qimage.cpp:41]: fair point actually.

Comment (LeonMatthesKDAB) [reply crates/cxx-qt-lib/src/gui/qimage.cpp]: Done

File: crates/cxx-qt-lib/src/core/qbytearray.rs (hunk 2 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-lib/src/core/qbytearray.rs]: Moved to #741

**Review by ahayzen-kdab — APPROVED** [review]
LGTM :-)

**Comment by ahayzen-kdab** [comment]
I'll rebase it after the other branch lands...
